Test Setup: Add more-complete mocks of common timer functions

dmsnell opened the pull request:
## Status

 - ~Created PR to watch tests pass, something locally was failing regarding `Docker`~
 - Ready for review

## Description

The existing mock of `requestAnimationFrame` omits a parameter when
calling the given callback. This patch adds that missing parameter.

Other code in the project also depends on `requestIdleCallback` but
this wasn't mocked at all, leading to test failures in a project
that integrates Gutenberg. This patch adds a reasonable mock of that
function as well.

These mocks have been rearranged as well so that their names match
the expected names by using the more verbose `= function request...`
syntax instead of directly setting their values to the `setTimeout`
functions.

## How has this been tested?

Please audit the code in here. This change should only impact the test
code and the `setup-globals.js` test configuration in the `jest-preset-default`
package. To fully test import the package in a project that runs tests on
code that imports `@wordpress/priority-queue`

In `trunk` the tests should fail with `TypeError: Cannot read property 'timeRemaining' of undefined`
In this branch the tests should pass.

## Types of changes

There should be no functional or visual changes in this fix.
Only test environments should be affected and with that
only currently-improperly-failing tests should start properly
passing.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

dmsnell committed:
Test Setup: Add more-complete mocks of common timer functions

The existing mock of `requestAnimationFrame` omits a parameter when
calling the given callback. This patch adds that missing parameter.

Other code in the project also depends on `requestIdleCallback` but
this wasn't mocked at all, leading to test failures in a project
that integrates Gutenberg. This patch adds a reasonable mock of that
function as well.

These mocks have been rearranged as well so that their names match
the expected names by using the more verbose `= function request...`
syntax instead of directly setting their values to the `setTimeout`
functions.

gziolo commented on packages/jest-preset-default/scripts/setup-globals.js:
```diff
@@ -5,8 +5,38 @@ global.window.tinyMCEPreInit = {
+		() =>
+			callback( {
+				didTimeout: false,
+				timeRemaining: () => Math.max( 0, 50 - ( Date.now() - start ) ),
```
Just confirming I follow the code, it returns a mocked value in the range 0-50.

dmsnell replied:
now that I reflect on this I'm not sure why that's there. I might have copied it from somewhere.

we should return _some_ value if `timeRemaining` is called. for the sake of the tests I don't think it will matter because the important part is that we return a non-negative number of `ms` remaining in the current idle period. for test purposes, this should be small if indeed anything were to rely on it, as you would not expect to get long windows of idle time in the browser

gziolo replied:
Right, I don't think it's something that you would want to use in the test for verification.

gziolo commented on packages/jest-preset-default/scripts/setup-globals.js:
```diff
@@ -5,8 +5,38 @@ global.window.tinyMCEPreInit = {
+	callback
+) {
+	// eslint-disable-next-line no-restricted-syntax
+	const randomDelay = Math.round( ( Math.random() * 1_000 ) / 60 );
```
Unrelated Aside: I'm wondering if the time period for animation frame becomes `1000 / 120` when the display has 120Hz like Apple devices with Pro Motion 😄

dmsnell replied:
this also shouldn't matter that much for the tests. by specification yes, we could get 120Hz > 8ms inter-frame times, but for the sake of mocking this function it will appear as if our fake display is 60Hz. code should not rely on the timing of the call but rather the `DOMHighResTimeStamp` passed into the callback if they want to perform time-based calculations.

gziolo replied:
Yes, it's irrelevant here what the number is here. I mostly found it intriguing that you often can read online that the animation frame is 16ms.

gziolo reviewed (approved):
Thank you for your contribution. That's definitely a better mock implementation 👍🏻 

We could add a note in the CHANGELOG for the package:
https://github.com/WordPress/gutenberg/blob/trunk/packages/jest-preset-default/CHANGELOG.md

dmsnell committed:
Add note in changelog

dmsnell commented:
Thanks @gziolo.

I added a note in the CHANGELOG but I'm not sure if I did it right. I added the note under `Unreleased`. If I need to assign a version number bump I can - I believe it would be 7.1.2 since this is a simple bug with that shouldn't break any existing code (unless of course they depended on the invalid behavior of `requestAnimationFrame`).

gziolo reviewed (approved):
All is good with the changelog. The script that performs npm publishing will inject a proper heading with the next version number and the date.
